File: April/30-4/loopLinklist.go
package main

import "fmt"

type Node struct {
	data int
	next *Node
}

type linkList struct {
	head *Node
}

func (ll *linkList) insert(val int) {
	new_node := &Node{data: val, next: nil}
	if ll.head == nil {
		ll.head = new_node
		return
	}
	tmp := ll.head
	for tmp != nil {
		tmp = tmp.next
	}
	tmp.next = new_node
}

func (ll *linkList) lengthLoop() int {
	slow := ll.head
	fast := ll.head
	f := false
	for fast != nil && fast.next != nil {
		slow = slow.next
		fast = fast.next.next
		if slow == fast {
			f = true
			break
		}
	}
	if !f {
		return 0
	}
	tmp := slow.next
	cnt := 1
	for tmp != nil {
		if slow == tmp {
			break
		}
		tmp = tmp.next
		cnt += 1
	}
	return cnt
}

func main() {
	ll := linkList{}
	ll.insert(1)
	ll.insert(2)
	fmt.Println(ll.lengthLoop())
}

File: April/30-4/loopLinklist.py
'''
Given the head of a linked list, determine whether the list contains a loop. If a loop is present, return the number of nodes in the loop, otherwise return 0.

Note: 'c' is the position of the node which is the next pointer of the last node of the linkedlist. If c is 0, then there is no loop.

Examples:

Input: head: 25 → 14 → 19 → 33 → 10 → 21 → 39 → 90 → 58 → 45, c = 4
Output: 7
Explanation: The loop is from 33 to 45. So length of loop is 33 → 10 → 21 → 39 → 90 → 58 → 45 = 7.
The number 33 is connected to the last node of the linkedlist to form the loop because according to the input the 4th node from the beginning(1 based indexing) 
will be connected to the last node in the LinkedList.

Input: head: 0 → 1 → 2 → 3, c = 0
Output: 0
Explanation: There is no loop.

Constraints:
1 ≤ no. of nodes ≤ 106
0 ≤ node.data ≤ 106
0 ≤ c ≤ n-1
'''

def countNodesInLoop(head):
    slow = head
    fast = head
    f = False
    while fast and fast.next:
        slow = slow.next
        fast = fast.next.next
        if slow == fast:
            f = True
            break
    if not f:
        return 0
    
    tmp = slow.next
    cnt = 1
    while tmp:
        if slow == tmp:
            return cnt
        tmp = tmp.next
        cnt += 1